### PR TITLE
Simplify MCM tests

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
MCM tests are slow and dominate the CI run time. New tests have been incorporated as new features came in without rationalizing the test suite as a whole. There are redundant or overly parametrized tests as a results.   

**Description of the Change:**
Combine a bunch of tests into `test_simple_dynamic_circuit` which tests all possible measurements are handle correctly by `dynamic_one_shot` post-processing. This needs not be parametrized over `reset` since it does not show up in post-processing.

Combine some tests into `test_multiple_measurements_and_reset` which checks that multiple measurements are handle correctly and that `reset` is properly applied while getting to the final state.

Combine all composite MCM measurements in `test_composite_mcms`.

`test_counts_return_type` remains as is. 

`test_broadcasting_qnode` has to do with transform processing, and hence there is no need to parametrize over `reset`.

**Benefits:**
Faster CI.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Review by mudit2812 — APPROVED** [review]
Thanks a bunch for working on this @vincentmr ! I'm quite happy with the changes you've made, and in my opinion the test quality is preserved.

One thing I've been thinking a lot about is that due to relying heavily on integration/end-to-end testing, we've skipped on unit tests for key functionality such as `apply_operation`'s dispatch for `MidMeasureMP` and `Conditional`. Those are the places where we should validate things like qubit reset, correct branch selection for `Conditional`, correct postselection behaviour, etc. However, I do not think that is necessary here.

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -48,9 +48,12 @@
 
 <h4>Mid-circuit measurements and dynamic circuits</h4>
 
+* Rationalize MCM tests.
```
Comment (AmintorDusko) [inline doc/releases/changelog-dev.md:51]: Could you please expand on that?

**Review by AmintorDusko — APPROVED** [review]
Nice job!

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -51,9 +51,14 @@
 
 <h4>Mid-circuit measurements and dynamic circuits</h4>
 
+* Rationalize MCM tests, removing most end-to-end tests from the native MCM test file,
+  but keeping one that validates multiple mid-circuit measurements with any allowed return
+  and interface end-to-end tests.
+  [(#5787)](https://github.com/PennyLaneAI/pennylane/pull/5787)
+
```
Comment (AmintorDusko) [inline doc/releases/changelog-dev.md:58]: :muscle:

**Pull request merged**